Update TTTrack for chi2 split

mdhildreth opened the pull request:
#### PR description:

Added capability of splitting chisq into separate xy and z calculations for different track fits. Added three new (as yet undefined) MVA-based track quality variables.  Cleaned up some of the code, including some variable assignments that were previously missing.

#### PR validation:

Track finding code runs with these new modifications.  TTTrack just stores information.

mdhildreth committed:
Fix several issues with new TTTrack class; add separate entries for splitting chi2 into xy and z

mdhildreth committed:
Fix missing edm dictionalry entries

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28793/13478

- This PR adds an extra 20KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28793/13478/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28793/13478/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

tomalin commented:
Some questions/comments:
1)	Can't we delete data member theEta, since we have theTanL? And why do we need theMomentum if we have thePhi, theTanL & theRInv? Converting between them on the fly (using calc at https://github.com/cms-sw/cmssw/blob/540bc00e40edb2d4f54c486dcbe8656fe3d858cf/DataFormats/L1TrackTrigger/interface/TTTrack.h#L229 ) presumably wouldn’t use much CPU. And they will waste disk space.
2)	All function & variables names should start with a lower case letter. e.g. trackSeed(),  numFitPars.
3)	Many functions are included for backwards compatibility. Now we have switched to CMSSW11, can’t we forget backwards compatiblity and just delete these? This would substantially shorten & simplify the code, and the redundant parameter nPar in the old functions is confusing. (Though we’d have to check if any code in CMSSW11 aside from L1Trk uses TTTrack). This affects getSector(), getPhiSector(), getEtaSector() & getPOCA(nPar), getRInv(nPar), getChi2(nPar), getChi2Red(nPar), getStubPtConsistentcy(nPar), setFitParNo(nPar). 
4)	Why do we need the old constructor that used chi2 instead of chi2(rphi) and chi2(rz)? – Is this just to support the pure tracklet algo?
5)	Why do we need the constructor TTTrack (aStubs)?  I can’t imagine when one would know the stubs on a track, but not the helix parameters.
6)	Add function stubPtConsistency(){return theStubPtConsistency;}
7)	Add function numFitPars(){return NumFitPars;}
8)	What does function trackSeed() return? If it is the seed type (i.e. layers used), then I suggest renaming it to trackSeedType().
9)	Function testTrackWordBits() doesn’t do anything?
10)	Can we delete Npars4=4 and Npars5=5? 
11)	Why do we need all the checks such as https://github.com/cms-sw/cmssw/blob/540bc00e40edb2d4f54c486dcbe8656fe3d858cf/DataFormats/L1TrackTrigger/interface/TTTrack.h#L303 that the number of fit params must be 4 or 5? Can’t we just delete them and shorten the code?
12)	The speed of light is inaccurate. Please change “MagConstant=0.3” to 0.299792458.
Cheers,
Ian

kpedro88 commented:
@mdhildreth please apply the code-format, otherwise the PR cannot be tested

skinnari commented on DataFormats/L1TrackTrigger/interface/TTTrack.h:
```diff
@@ -99,11 +121,22 @@ class TTTrack : public TTTrack_TrackWord {
+  void setEtaSector(unsigned int aSector) { theEtaSector = aSector; }
+
   /// Track seeding (for debugging)
   unsigned int TrackSeed() const { return theTrackSeed; }
```
this should be trackSeed() for consistency

skinnari commented on DataFormats/L1TrackTrigger/interface/TTTrack.h:
```diff
@@ -35,12 +35,16 @@ class TTTrack : public TTTrack_TrackWord {
   unsigned int theHitPattern;
-  float an_MVA_Value;
-  float another_MVA_Value;
+  double theTrkMVA1;
```
do we want these to be doubles or unsigned ints?

skinnari commented:
@mdhildreth @tomalin important bug still, this is wrong: 

https://github.com/mdhildreth/cmssw/blob/uptdateTTTrackForChi2Split/DataFormats/L1TrackTrigger/interface/TTTrack.h#L228

that is a calculation that uses radius of curvature in the formula, but the variable is inverse radius of curvature! must also clarify in the code that the assumed unit is [m-1]. at least the tracklet uses by default internally [cm-1] and then does the conversion when calculating pt, so would be good to add that as a comment in the class.

mdhildreth committed:
implement suggested changes, fix momentum calculation

mdhildreth committed:
code format checks

cmsbuild commented:
The code-checks are being triggered in jenkins.

mdhildreth commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28793/13641

- This PR adds an extra 24KB to repository

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
A new Pull Request was created by  @mdhildreth  for master.

It involves the following packages:

DataFormats/L1TrackTrigger


@cmsbuild, @rekovic, @benkrikler, @kpedro88 can you please review it and eventually sign? Thanks.
@rovere, @sviret this is something you requested to watch as well.
@davidlange6, @silviodonato, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

cmsbuild commented:
-1

Tested at: 5eca77b10e72a14e136cfea617535c2c5d27ef2b

CMSSW: CMSSW_11_1_X_2020-02-05-1100
SCRAM_ARCH: slc7_amd64_gcc820
You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a83927/4515/summary.html

I found follow errors while testing this PR

**Failed tests:  Build**

* **Build**:

I found compilation error when building: 

<pre>>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_1_X_2020-02-05-1100/src/L1Trigger/TrackTrigger/src/TTClusterAlgorithm_official.cc 
>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_1_X_2020-02-05-1100/src/L1Trigger/TrackTrigger/src/TTStubAlgorithm_official.cc 
>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_1_X_2020-02-05-1100/src/L1Trigger/TrackTrigger/src/TTStubAlgorithm_cbc3.cc 
>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_1_X_2020-02-05-1100/src/L1Trigger/TrackTrigger/src/TTClusterAlgorithm_neighbor.cc 
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_1_X_2020-02-05-1100/src/L1Trigger/TrackTrigger/src/StubPtConsistency.cc: In function 'float StubPtConsistency::getConsistency(TTTrack<edm::Ref<edm::DetSetVector<Phase2TrackerDigi>, Phase2TrackerDigi> >, const TrackerGeometry*, const TrackerTopology*, double, int)':
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_1_X_2020-02-05-1100/src/L1Trigger/TrackTrigger/src/StubPtConsistency.cc:24:82: error: 'class TTTrack<edm::Ref<edm::DetSetVector<Phase2TrackerDigi>, Phase2TrackerDigi> >' has no member named 'getRInv'; did you mean 'theRInv'?
     float trk_signedPt = speedOfLightConverted * mMagneticFieldStrength / aTrack.getRInv(nPar);
                                                                                  ^~~~~~~
                                                                                  theRInv
gmake: *** [tmp/slc7_amd64_gcc820/src/L1Trigger/TrackTrigger/src/L1TriggerTrackTrigger/StubPtConsistency.cc.o] Error 1
>> Building shared library tmp/slc7_amd64_gcc820/src/L1Trigger/TrackTrigger/src/L1TriggerTrackTrigger/libL1TriggerTrackTrigger.so
</pre>

cmsbuild commented:
Comparison not run due to Build errors (RelVals and Igprof tests were also skipped)

mdhildreth committed:
Yeah, checkdeps...

mdhildreth commented:
please test

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28793/13648

- This PR adds an extra 40KB to repository

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #28793 was updated. @benkrikler, @kmaeshima, @schneiml, @andrius-k, @kpedro88, @cmsbuild, @rekovic, @jfernan2, @fioriNTU can you please check and sign again.

cmsbuild commented:
+1
Tested at: fcdd9ea5a45680d3d642700974a01bea46a060b0
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a83927/4529/summary.html
CMSSW: CMSSW_11_1_X_2020-02-05-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a83927/4529/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2698043
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2697696
 - DQMHistoTests: Total skipped: 346
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

jfernan2 commented:
+1

tomalin commented:
A few remaining comments:
1) Constructor https://github.com/cms-sw/cmssw/blob/fcdd9ea5a45680d3d642700974a01bea46a060b0/DataFormats/L1TrackTrigger/interface/TTTrack.h#L165 should initialize theChi2XY & theChi2Z to zero.
2) Can't we delete the constructor https://github.com/cms-sw/cmssw/blob/fcdd9ea5a45680d3d642700974a01bea46a060b0/DataFormats/L1TrackTrigger/interface/TTTrack.h#L185 ? When would we know stubs but not helix params?
3) Constructor https://github.com/cms-sw/cmssw/blob/fcdd9ea5a45680d3d642700974a01bea46a060b0/DataFormats/L1TrackTrigger/interface/TTTrack.h#L240 should initialize theChi2=theChi2XY+theChi2Z;
4) Constructor https://github.com/cms-sw/cmssw/blob/fcdd9ea5a45680d3d642700974a01bea46a060b0/DataFormats/L1TrackTrigger/interface/TTTrack.h#L240 should initialize theChi2XY & theChi2Z to some value to indicate not set.
5) Add function chi2XY(). Suggest renaming chi2z() to chi2Z(). And add function dof() to return (2 * theStubRefs.size() - numFitPars); -- it can be called also from chi2Red().
6) Suggest setting B-field value via setBfield() function instead of constructor. In data, best estimate of B-field may change after dataset is produced. This would require calculating Pt on the fly instead of in constructor. (Alternative if so keep B-field in constructor, but provide setBfield() to set B and update thePt and theMomentum).
7) TrackWord should take chi2XY and chi2Z (corresponding to firmware), not chi2Z and chi2.

skinnari commented:
@mdhildreth this class is still buggy. please do not merge yet. 

(1) this: 

https://github.com/cms-sw/cmssw/blob/fcdd9ea5a45680d3d642700974a01bea46a060b0/DataFormats/L1TrackTrigger/interface/TTTrack.h#L219

should be sinh(aTanlambda) 

(2) neither theD0 or theZ0 are set in the constructors.

mdhildreth committed:
a couple more fixes

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28793/13659

- This PR adds an extra 44KB to repository

cmsbuild commented:
Pull request #28793 was updated. @benkrikler, @kmaeshima, @schneiml, @andrius-k, @kpedro88, @cmsbuild, @rekovic, @jfernan2, @fioriNTU can you please check and sign again.

mdhildreth commented:
@kpedro88 @rekovic This PR is ready to go now.  If you have additional comments, please go ahead and make them.  Thanks.

skinnari commented:
> A few remaining comments:
> 
>     1. Constructor https://github.com/cms-sw/cmssw/blob/fcdd9ea5a45680d3d642700974a01bea46a060b0/DataFormats/L1TrackTrigger/interface/TTTrack.h#L165
>         should initialize theChi2XY & theChi2Z to zero.
> 
>     2. Can't we delete the constructor https://github.com/cms-sw/cmssw/blob/fcdd9ea5a45680d3d642700974a01bea46a060b0/DataFormats/L1TrackTrigger/interface/TTTrack.h#L185
>         ? When would we know stubs but not helix params?
> 
>     3. Constructor https://github.com/cms-sw/cmssw/blob/fcdd9ea5a45680d3d642700974a01bea46a060b0/DataFormats/L1TrackTrigger/interface/TTTrack.h#L240
>         should initialize theChi2=theChi2XY+theChi2Z;
> 
>     4. Constructor https://github.com/cms-sw/cmssw/blob/fcdd9ea5a45680d3d642700974a01bea46a060b0/DataFormats/L1TrackTrigger/interface/TTTrack.h#L240
>         should initialize theChi2XY & theChi2Z to some value to indicate not set.
> 
>     5. Add function chi2XY(). Suggest renaming chi2z() to chi2Z(). And add function dof() to return (2 * theStubRefs.size() - numFitPars); -- it can be called also from chi2Red().
> 
>     6. Suggest setting B-field value via setBfield() function instead of constructor. In data, best estimate of B-field may change after dataset is produced. This would require calculating Pt on the fly instead of in constructor. (Alternative if so keep B-field in constructor, but provide setBfield() to set B and update thePt and theMomentum).
> 
>     7. TrackWord should take chi2XY and chi2Z (corresponding to firmware), not chi2Z and chi2.

@mdhildreth were these addressed? in particular there are some initializations that were missing?

skinnari commented:
just to note also here, it seems get_ichi2() results in seg fault, and that the TTTrack word does not produce the expected results. @mdhildreth

kpedro88 commented on DataFormats/L1TrackTrigger/interface/TTTrack.h:
```diff
@@ -31,21 +31,24 @@ class TTTrack : public TTTrack_TrackWord {
-  unsigned int NumFitPars;
+  double theChi2XY;
+  double theChi2Z;
+  unsigned int numFitPars;
```
to fit the style used in this class, should be named `theNumFitPars`
(to fit CMS style, all variables should be changed to have a trailing underscore...)

mdhildreth replied:
done.

kpedro88 commented on DataFormats/L1TrackTrigger/interface/TTTrack.h:
```diff
@@ -31,21 +31,24 @@ class TTTrack : public TTTrack_TrackWord {
   static constexpr unsigned int Npars4 = 4;
   static constexpr unsigned int Npars5 = 5;
-  static constexpr float MagConstant = 0.3;
+  static constexpr float MagConstant = 0.299792458;
```
better to define this in terms of `CLHEP::c_light`?

mdhildreth replied:
done

kpedro88 commented on DataFormats/L1TrackTrigger/interface/TTTrack.h:
```diff
@@ -156,34 +189,92 @@ TTTrack<T>::TTTrack(std::vector<edm::Ref<edmNew::DetSetVector<TTStub<T> >, TTStu
   theStubRefs.clear();
-  double thePT = MagConstant * aRinv * aBfield;
-  theMomentum = GlobalVector(GlobalVector::Cylindrical(thePT, aphi, thePT * sinh(aeta)));
+  double thePT = fabs(MagConstant * 1.0 / aRinv * aBfield / 100.0);  // Rinv is in cm-1
```
use `std::abs()` rather than `fabs()`

mdhildreth replied:
done

kpedro88 commented on DataFormats/L1TrackTrigger/interface/TTTrack.h:
```diff
@@ -156,34 +189,92 @@ TTTrack<T>::TTTrack(std::vector<edm::Ref<edmNew::DetSetVector<TTStub<T> >, TTStu
+  numFitPars = nPar;
+  theHitPattern = aHitPattern;
+  theBField = aBfield;
+  // should probably fill the momentum vectur
```
is this needed for this PR?

mdhildreth replied:
was just a random comment. It's done.

kpedro88 commented on DataFormats/L1TrackTrigger/interface/TTTrack.h:
```diff
@@ -156,34 +189,92 @@ TTTrack<T>::TTTrack(std::vector<edm::Ref<edmNew::DetSetVector<TTStub<T> >, TTStu
+                    unsigned int nPar,
+                    double aBfield) {
+  theStubRefs.clear();
+  double thePT = fabs(MagConstant * 1.0 / aRinv * aBfield / 100.0);  // Rinv is in cm-1
```
use `std::abs()` rather than `fabs()`

mdhildreth replied:
done

kpedro88 commented on DataFormats/L1TrackTrigger/interface/TTTrack.h:
```diff
@@ -156,34 +189,92 @@ TTTrack<T>::TTTrack(std::vector<edm::Ref<edmNew::DetSetVector<TTStub<T> >, TTStu
+  // should probably fill the momentum vectur
+}
+
+/// Second default constructor with split chi2
```
rather than duplicating all this code, call the previous constructor and then just change the chi2 values?

mdhildreth replied:
done

kpedro88 commented on DataFormats/L1TrackTrigger/interface/TTTrack.h:
```diff
@@ -205,137 +296,96 @@ void TTTrack<T>::setFitParNo(unsigned int nPar) {
-    return theChi2 / (2 * theStubRefs.size() - NumFitPars);
-  } else
-    return 0.0;
+  return theChi2 / (2 * theStubRefs.size() - numFitPars);
```
is it worth caching this value (maybe as a transient to maintain the size of the class on disk)?

mdhildreth replied:
@kpedro88 Wouldn't caching make the class bigger?  I don't understand the sign of this comment?

kpedro88 replied:
Member variables can be marked as transient so they do not persist, but also do not need to be recalculated every time the accessor is called.

kpedro88 commented on DataFormats/L1TrackTrigger/src/TTTrack_TrackWord.cc:
```diff
@@ -297,8 +314,19 @@ unsigned int TTTrack_TrackWord::get_BendChi2Bits() {
 }
 
+float TTTrack_TrackWord::unpack_ichi2Z() {
+  unsigned int bits = (TrackWord3 & 0x00003C00) >> 10;
```
all of the masks and offsets in this file should be named constants

skinnari commented:
@mdhildreth In addition to what has been pointed out above (e.g. related to chi2 from @tomalin) and general comments from @kpedro88), also a getter function for hitpattern is missing!

mdhildreth commented:
back out pushing entire L1 Trigger suite.  Oops.  More later.

mdhildreth committed:
add a few new features, address almost all review comments

cmsbuild commented:
The code-checks are being triggered in jenkins.

mdhildreth commented:
@tomalin @skinnari  This last commit addresses all of the comments and suggestions except it doesn't
1) fix the get_ichi2() bug/crash
2) rename all of the bit patterns in the TrackWord as variables.  Too much work for now.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28793/13848

- This PR adds an extra 44KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28793/13848/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28793/13848/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

skinnari commented on DataFormats/L1TrackTrigger/interface/TTTrack.h:
```diff
@@ -63,7 +63,7 @@ class TTTrack : public TTTrack_TrackWord {
           double trkMVA1,
           double trkMVA2,
-          double trkMVA3,
+          double trkMVA3_,
```
@mdhildreth why did trkMVA3 change to trkMVA3_ in the two constructors? This seems odd to me?

mdhildreth replied:
oh, that's a typo.  Shouldn't be...  It doesn't really matter, but...

skinnari commented on DataFormats/L1TrackTrigger/interface/TTTrack.h:
```diff
@@ -26,34 +27,33 @@ class TTTrack : public TTTrack_TrackWord {
+  unsigned int theEtaSector_;
+  double theStubPtConsistency_;
+  double theChi2_;
+  double theChi2_XY_;
```
Were these meant to be theChi2XY_ and theChi2Z_ ?

skinnari commented on DataFormats/L1TrackTrigger/interface/TTTrack.h:
```diff
@@ -296,96 +280,119 @@ void TTTrack<T>::setFitParNo(unsigned int nPar) {
+  return theStubPtConsistency_;
 }
 
+/// StubPtConsistency
```
Would be good to correct comment to avoid confusion...

mdhildreth replied:
done

skinnari commented on DataFormats/L1TrackTrigger/interface/TTTrack.h:
```diff
@@ -296,96 +280,119 @@ void TTTrack<T>::setFitParNo(unsigned int nPar) {
+  return theHitPattern_;
+}
+
+/// StubPtConsistency
```
Same here, would be good to correct comment to avoid confusion.

tomalin replied:
Agreed. I have a few further comments:

1) The digitisation ranges assumed at https://github.com/mdhildreth/cmssw/blob/uptdateTTTrackForChi2Split/DataFormats/L1TrackTrigger/interface/TTTrack_TrackWord.h#L219 appear wildly inconsistent with those specified in https://twiki.cern.ch/twiki/bin/view/CMS/HybridDataFormat#Fitted_Tracks_written_by_KalmanF . Please check.

2) Please add two extra functions to TTTrack.h:

chi2XYred(){return theChi2XY_/(theStubRefs.size() - (theNumFitPars_-2));}
chi2Zred(){return theChi2Z_/(theStubRefs.size() - 2);}

3) If you have adopted the convention that class data members are distinguished by naming them theXYZ_, then variables such as https://github.com/mdhildreth/cmssw/blob/uptdateTTTrackForChi2Split/DataFormats/L1TrackTrigger/interface/TTTrack_TrackWord.h#L26 should not include "the"

4) Please eliminate the hard-wired constants, such as 16 at https://github.com/mdhildreth/cmssw/blob/uptdateTTTrackForChi2Split/DataFormats/L1TrackTrigger/src/TTTrack_TrackWord.cc#L167 . This could be written instead (NZ0Bits+1)+NChi2Bits .

Cheers,
Ian

mdhildreth replied:
**Responding to comments:**

The last commit gets most of these, including Louise's suggestions.

Responding to Ian's comments: @tomalin 

1. Fixed, except for phi0. I assumed that phi would be digitized with respect to the center of a nonant, which gets you maximum resolution (if we could figure out how to calculate it).  I don't know/understand why phi0 should range between +/- 1.026 ?  Why is that the magic number?

2. Done
3. Done
4. Sorry, I don't have time to do this now.  Since the Track Word format is likely to change, I also don't want to spend a whole bunch of time prettying up this code at this point.  I suggest we press on with this version unless it's critical.  
5. Done

Cheers,
Mike

tomalin replied:
1) The digitisation range in phi0 of +-1.026 rads about the centre of the sector. This is because the required stub phi digi range is 2*pi/9 + 2*hourglass_overlap (2*0.1638) = 1.026; i.e. from -1.026/2 to +1.026/2. To simplify maths, the phi0 digi range can differ from this only by powers of 2. And unfortunately, this range is very slightly too small to cover the allowed phi0 values of tracks in a nonant, so we have to double it to (-1.026,+1.026).
2) I just noticed that TTTrackWord is digitising eta instead of tanL. This is also inconsistent with the agreement documented on the twiki.

mdhildreth replied:
@tomalin @skinnari
Ok, I have fixed the last two issues with the phi digitization and switched eta -> tanl in the Track Word.   I guess someone downstream will do the conversion from tanl back to eta so it can be used in the trigger.

tomalin replied:
My comment (2) above was incorrectly implemented in the C++. The number of degrees of freedom of the chi2XY should be "(theStubRefs.size() - (theNumFitPars_-2))", but at https://github.com/cms-sw/cmssw/blob/9deefb0d87de9ad4a85b9c12cf28723725636493/DataFormats/L1TrackTrigger/interface/TTTrack.h#L347 it is instead given as "(theStubRefs.size() - theNumFitPars_ - 2)". 

N.B. As sanity check, the number of degrees of freedom in XY plus the number in Z should equal the number in total, which is not currently the case in the C++.

mdhildreth replied:
fixed

mdhildreth committed:
rebase, fix some typos

mdhildreth committed:
code formatting

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28793/13946

- This PR adds an extra 44KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/L1TrackTrigger/src/classes_def.xml  modified in PR(s): #29038

cmsbuild commented:
Pull request #28793 was updated. @benkrikler, @kmaeshima, @schneiml, @andrius-k, @kpedro88, @cmsbuild, @rekovic, @jfernan2, @fioriNTU can you please check and sign again.

mdhildreth committed:
More cleanups, renaming, responding to comments

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28793/13947

- This PR adds an extra 44KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/L1TrackTrigger/src/classes_def.xml  modified in PR(s): #29038

cmsbuild commented:
Pull request #28793 was updated. @benkrikler, @kmaeshima, @schneiml, @andrius-k, @kpedro88, @cmsbuild, @rekovic, @jfernan2, @fioriNTU can you please check and sign again.

mdhildreth committed:
Last cleanup, hopefully

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28793/13957

- This PR adds an extra 24KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/L1TrackTrigger/src/classes_def.xml  modified in PR(s): #29038

cmsbuild commented:
Pull request #28793 was updated. @benkrikler, @kmaeshima, @schneiml, @andrius-k, @kpedro88, @cmsbuild, @rekovic, @jfernan2, @fioriNTU can you please check and sign again.

silviodonato commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 5fc8ea1c16a3dfde0f8f7437338edc7eceeedac0
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a83927/4964/summary.html
CMSSW: CMSSW_11_1_X_2020-03-03-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a83927/4964/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 6 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2680373
 - DQMHistoTests: Total failures: 41
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2680013
 - DQMHistoTests: Total skipped: 319
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

jfernan2 commented:
+1
Changes unrelated to this PR

kpedro88 commented:
https://github.com/cms-sw/cmssw/pull/28793/files#r378293112 has not been addressed, leaving several hardcoded values in TTTrack_TrackWord.

mdhildreth committed:
final updates plus code checks

cmsbuild commented:
The code-checks are being triggered in jenkins.

mdhildreth commented:
@kpedro88 Hi Kevin - I updated the fixed masks and shifts in TTTrack_TrackWord.cc  Everything else should be fixed now.

mdhildreth commented:
please test

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28793/14166

- This PR adds an extra 92KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/L1TrackTrigger/interface/TTTypes.h  modified in PR(s): #29151
  - File  DataFormats/L1TrackTrigger/src/classes_def.xml  modified in PR(s): #29038, #29151


Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28793/14166/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28793/14166/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

mdhildreth committed:
code formatting - I really am tired of this

mdhildreth commented:
please test

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28793/14167

- This PR adds an extra 84KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/L1TrackTrigger/interface/TTTypes.h  modified in PR(s): #29151
  - File  DataFormats/L1TrackTrigger/src/classes_def.xml  modified in PR(s): #29038, #29151

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #28793 was updated. @benkrikler, @kmaeshima, @schneiml, @andrius-k, @kpedro88, @cmsbuild, @rekovic, @jfernan2, @fioriNTU can you please check and sign again.

cmsbuild commented:
+1
Tested at: 9deefb0d87de9ad4a85b9c12cf28723725636493
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a83927/5149/summary.html
CMSSW: CMSSW_11_1_X_2020-03-11-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a83927/5149/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2680593
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2680273
 - DQMHistoTests: Total skipped: 319
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

jfernan2 commented:
+1

kpedro88 commented:
+upgrade

silviodonato commented:
@benkrikler @rekovic do you have any comments?

tomalin commented:
Please check my report of a bug in TTTrack.h in https://github.com/cms-sw/cmssw/pull/28793#discussion_r392960098 . (I mention it again here, as github "hides" old 
conversations, and I suspect it could be missed).

silviodonato commented on DataFormats/L1TrackTrigger/interface/TTTrack.h:
```diff
@@ -205,137 +279,130 @@ void TTTrack<T>::setFitParNo(unsigned int nPar) {
+/// Chi2XY reduced
+template <typename T>
+double TTTrack<T>::chi2XYRed() const {
+  return theChi2_XY_ / (theStubRefs.size() - theNumFitPars_ - 2);
```
```suggestion
  return theChi2_XY_ / (theStubRefs.size() - (theNumFitPars_ - 2));
```
@mdhildreth, @tomalin suggest to make this fix

mdhildreth committed:
Update DataFormats/L1TrackTrigger/interface/TTTrack.h

Co-Authored-By: Silvio Donato <silvio.donato@cern.ch>

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28793/14213

- This PR adds an extra 92KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/L1TrackTrigger/interface/TTTypes.h  modified in PR(s): #29151
  - File  DataFormats/L1TrackTrigger/src/classes_def.xml  modified in PR(s): #29038, #29151

cmsbuild commented:
Pull request #28793 was updated. @benkrikler, @kmaeshima, @schneiml, @andrius-k, @kpedro88, @cmsbuild, @rekovic, @jfernan2, @fioriNTU can you please check and sign again.

rekovic commented:
+1

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 2abefd24bce74af3e18858048ac68eb9d6a8c559
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a83927/5212/summary.html
CMSSW: CMSSW_11_1_X_2020-03-16-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a83927/5212/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2680593
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2680273
 - DQMHistoTests: Total skipped: 319
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

jfernan2 commented:
+1

kpedro88 commented:
+upgrade

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @silviodonato, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

silviodonato commented:
+1
